addons/print/models/print_printer.py

```diff
@@ -169,34 +185,49 @@ def spool_report(self, docids, report_name, data=None, title=None,
             reports = report_name
         else:
             name = report_name
-            Report = self.env['ir.actions.report']
+            Report = self.env["ir.actions.report"]
             reports = Report._get_report_from_name(name)
             if not reports:
                 reports = self.env.ref(name, raise_if_not_found=False)
             if not reports:
                 raise UserError(_("Undefined report %s") % name)
 
         # Identify required report types
-        printers = self.printers(raise_if_not_found=True)
-        required = set(printers.mapped('report_type'))
-        available = set(reports.mapped('report_type'))
-        missing = required - available
+        report_types = set(reports.mapped("report_type"))
+
+        # Identify printer to use for each report type, only include report type in dictionary
+        # if a printer is identified
+        printers_by_report_type = {
+            rt: p for rt in report_types for p in (self.printers(report_type=rt)) if p
+        }
+
+        if printers_by_report_type:
+            required_types = set(printers_by_report_type.keys())
+            missing = required_types - report_types
```

Comment:
Should this be the other way around? `missing` should be the types that we require, less the types we have. Here `required_types` looks to be the types we have. It's worth noting that I think the previous `required` and `available` variables are the wrong way round.

Reply:
Might be that the unit tests updating, to explain why it is setup like this at the moment:

`test01_spool_test_page` should print one test page - https://github.com/unipartdigital/odoo-print/blob/story/10476/addons/print/tests/test_print_printer.py#L104

When it is run the variables here are as follows:

`reports`: 'Printer Test Page (CPCL)', 'Printer Test Page (PDF)' (recordset of `ir.actions.report`)
`required_types`: {'qweb-pdf'}
`report_types`: {'qweb-cpcl', 'qweb-pdf'}

So if I did it the other way around it would complain that the CPCL report couldn't be printed. But if I remove this check completely, the unit test fails because it was only expected the PDF report to be printed.

Maybe the logic for some of the unit tests or `spool_test_page` is a bit backwards also?

Reply:
To add to this, `test_page_report` currently returns a PDF and a CPCL test report.

This is the commit that added in the function: https://github.com/unipartdigital/odoo-print/commit/f9ea73ffb18dc4342aaa5e10f9ca45a0f5bd18df

Not sure how best to proceed, I agree the logic does seem a bit backwards but also worried about breaking existing functionality!

Reply:
We were discussing with @joe-gb about this the other day and the key is that we are in the printer model, but when we think about it we have in mind that we need a printer to print a given report(s). That is why the logic is backwards, we have a list of printers and we have to find a list of reports that match them, instead of list of reports and we have to find printers to print them.

Reply:
Have talked with @apatrick88, as long as the changes in the story work as expected we'll raise a separate issue to look into this further.